examples/chain_replication.pyv

```diff
@@ -0,0 +1,89 @@
+# node ids and states
+sort node_id
+sort node_state
+
+# ordering on node states
+immutable relation le(node_state, node_state)
+axiom le(X,X) # reflexive
+axiom le(X,Y) & le(Y,X) -> X = Y # anti-symmetric
+axiom le(X,Y) & le(Y,Z) -> le(X,Z) # transitive
+axiom le(X,Y) | le(Y,X) # total
+
+# empty state
+immutable constant empty: node_state
+
+# the empty state is the least state
+axiom le(empty, X)
+
+# let's say that there are three nodes
+immutable constant head: node_id
+immutable constant middle: node_id
+immutable constant tail: node_id
+
+# the head, middle, and tail nodes are all different
+# (this is too strong; it should be OK for them to all be the same, too)
+axiom head != middle & head != tail & middle != tail
+
+# nodes have states
+# TODO: maybe this could just be a big mutable relation
+mutable constant head_state : node_state
+mutable constant middle_state : node_state
+mutable constant tail_state : node_state
+
+# set of messages in transit: sender, recipient, data
+mutable relation msg(node_id, node_id, node_state)
+
+# all replicas start in the empty state
+init head_state = empty
+init middle_state = empty
+init tail_state = empty
+# with no messages in transit
+init !msg(X, Y, Z)
+
+# the head replica can move to a new state and send its state to the middle node
+transition head_send(next: node_state)
+  modifies head_state, msg
+  le(head_state, next) & # the new state must be bigger
+  new(head_state) = next &
+  # add next to the set of messages sent from head to middle
+  (new(msg(X, Y, Z)) <-> msg(X, Y, Z) | (X = head & Y = middle & Z = next))
```

Comment:
Another option here would be to split this into two transitions: one where the head changes its state to next, and a second where head sends whatever its current state is to middle. The reason I mention it is that there is a discrepancy between head's behavior and middle's behavior. The middle node splits the operations of receiving the new state and sending it on, while head has these two operations joined atomically.

Reply:
Yeah, the discrepancy bothered me too.  Also, really, instead of messages originating at head, there should be some `client` node that produces updates and sends them to head.  Then the code for head would be identical to that for middle, and the client can have distinct `client_update` and `client_send` transitions for producing and sending a new state.